### DISCUSSION TITLE
Can Ocelot support SSO and JWT authentication both at same time?

### DISCUSSION THREAD

**Discussion by HorstWang** [ThreeMammals/Ocelot#2013] [Q&A] [answered]
Hello,

I have added a JWT authentication to my Ocelot gateway to verify JWT token issued by Azure Active Directory and it works fine.
But I want to do even more to add an OpenID Connect integration with Azure Active Directory, so that Ocelot can redirect to login for unauthenticated users and automatically acquire access token on custom scope and maintain this token in user's session and refresh it when needed, so that we can completely off load authentication code out of our applications to Ocelot gateway.
I don't know if this is supported on Ocelot.
Would you give us some suggestion?

**Comment by gao-artur** [comment]
I think this approach should solve your problem:

https://stackoverflow.com/questions/61020375/using-multiple-authorization-providers-cookie-auth-and-open-id-connect/61029461#61029461

You should use `"AuthenticationProviderKey" : "dynamic"` in ocelot.json

**Reply by raman-m** [reply]
Yes, defining authentication policy is an option. And this is native asp.net feature.
We have another Ocelot's feature, see ->
- [Multiple Authentication Schemes](https://ocelot.readthedocs.io/en/latest/features/authentication.html#multiple-authentication-schemes)

**Comment by raman-m** [comment, marked as answer]
@HorstWang 
Hi Horst!
Use Ocelot feature
- [Multiple Authentication Schemes](https://ocelot.readthedocs.io/en/latest/features/authentication.html#multiple-authentication-schemes)

This feature is available starting version [23.0](https://github.com/ThreeMammals/Ocelot/releases/tag/23.0.0)
**FYI** You could develop a solution based on ASP.NET native feature called Authentication Policies.
